clippy_lints/src/unintentional_unit_return.rs

```diff
@@ -0,0 +1,176 @@
+use crate::utils::{get_trait_def_id, paths, span_lint, span_lint_and_help};
+use if_chain::if_chain;
+use rustc_hir::def_id::DefId;
+use rustc_hir::{Expr, ExprKind, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_middle::ty::{GenericPredicates, PredicateKind, ProjectionPredicate, TraitPredicate};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{BytePos, Span};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for functions that expect closures of type
+    /// Fn(...) -> Ord where the implemented closure has a semi-colon
+    /// at the end of the last statement.
+    ///
+    /// **Why is this bad?** Likely the semi-colon is unintentional which
+    /// returns () instead of the result of last statement. Since () implements Ord
+    /// it doesn't cause a compilation error
+    ///
+    /// **Known problems:** If returning unit is intentional, then there is no
+    /// way of specifying this without triggering needless_return lint
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let mut twins = vec!((1,1), (2,2));
+    /// twins.sort_by_key(|x| { x.1; });
+    /// ```
+    pub UNINTENTIONAL_UNIT_RETURN,
+    nursery,
```

Comment:
I think we can move this to `correctness` since it should be finished now. Or do you know of any problems you encountered with this lint, while implementing/testing it?

Reply:
Nope, I think it serves the exact purpose it was made for.